File: frontend/lib/features/profile/data/models/user_profile_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'user_profile_model.g.dart';

/// Gender enum
enum Gender {
  @JsonValue('male')
  male,
  @JsonValue('female')
  female,
  @JsonValue('other')
  other,
  @JsonValue('prefer_not_to_say')
  preferNotToSay,
}

/// Fitness goal enum
enum FitnessGoal {
  @JsonValue('strength')
  strength,
  @JsonValue('muscle_gain')
  muscleGain,
  @JsonValue('fat_loss')
  fatLoss,
  @JsonValue('endurance')
  endurance,
  @JsonValue('general_fitness')
  generalFitness,
}

/// Experience level enum
enum ExperienceLevel {
  @JsonValue('beginner')
  beginner,
  @JsonValue('intermediate')
  intermediate,
  @JsonValue('advanced')
  advanced,
}

/// Nutrition goal enum
enum NutritionGoal {
  @JsonValue('cut')
  cut,
  @JsonValue('bulk')
  bulk,
  @JsonValue('recomp')
  recomp,
  @JsonValue('maintain')
  maintain,
}

/// User profile model
@JsonSerializable()
class UserProfileModel {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  final String email;

  // Basic info
  @JsonKey(name: 'full_name')
  final String? fullName;
  @JsonKey(name: 'phone_number')
  final String? phoneNumber;
  final Gender? gender;

  // Physical attributes
  @JsonKey(name: 'weight_kg')
  final double? weightKg;
  @JsonKey(name: 'height_cm')
  final double? heightCm;

  // Fitness preferences
  @JsonKey(name: 'fitness_goal')
  final FitnessGoal? fitnessGoal;
  @JsonKey(name: 'experience_level')
  final ExperienceLevel? experienceLevel;
  @JsonKey(name: 'training_frequency')
  final int? trainingFrequency;
  @JsonKey(name: 'nutrition_goal')
  final NutritionGoal? nutritionGoal;

  UserProfileModel({
    required this.id,
    required this.userId,
    required this.email,
    this.fullName,
    this.phoneNumber,
    this.gender,
    this.weightKg,
    this.heightCm,
    this.fitnessGoal,
    this.experienceLevel,
    this.trainingFrequency,
    this.nutritionGoal,
  });

  factory UserProfileModel.fromJson(Map<String, dynamic> json) {
    // Create a mutable copy so we can normalise incoming payload discrepancies.
    final normalised = Map<String, dynamic>.from(json);

    final resolvedUserId = _resolveUserId(json);
    normalised['user_id'] = resolvedUserId;

    // Some responses only include `name`; prefer backend value when available.
    normalised['full_name'] = normalised['full_name'] ?? normalised['name'];

    return _$UserProfileModelFromJson(normalised);
  }

  static String _resolveUserId(Map<String, dynamic> json) {
    final dynamic raw = json['user_id'] ?? json['userId'] ?? json['id'];
    if (raw is String && raw.isNotEmpty) {
      return raw;
    }
    throw const FormatException('Profile payload missing required user identifier');
  }

  Map<String, dynamic> toJson() => _$UserProfileModelToJson(this);

  UserProfileModel copyWith({
    String? id,
    String? userId,
    String? email,
    String? fullName,
    String? phoneNumber,
    Gender? gender,
    double? weightKg,
    double? heightCm,
    FitnessGoal? fitnessGoal,
    ExperienceLevel? experienceLevel,
    int? trainingFrequency,
    NutritionGoal? nutritionGoal,
  }) {
    return UserProfileModel(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      email: email ?? this.email,
      fullName: fullName ?? this.fullName,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      gender: gender ?? this.gender,
      weightKg: weightKg ?? this.weightKg,
      heightCm: heightCm ?? this.heightCm,
      fitnessGoal: fitnessGoal ?? this.fitnessGoal,
      experienceLevel: experienceLevel ?? this.experienceLevel,
      trainingFrequency: trainingFrequency ?? this.trainingFrequency,
      nutritionGoal: nutritionGoal ?? this.nutritionGoal,
    );
  }
}

/// Extension methods for enums to get display strings
extension GenderExtension on Gender {
  String get displayName {
    switch (this) {
      case Gender.male:
        return 'Male';
      case Gender.female:
        return 'Female';
      case Gender.other:
        return 'Other';
      case Gender.preferNotToSay:
        return 'Prefer not to say';
    }
  }
}

extension FitnessGoalExtension on FitnessGoal {
  String get displayName {
    switch (this) {
      case FitnessGoal.strength:
        return 'Strength';
      case FitnessGoal.muscleGain:
        return 'Muscle Gain';
      case FitnessGoal.fatLoss:
        return 'Fat Loss';
      case FitnessGoal.endurance:
        return 'Endurance';
      case FitnessGoal.generalFitness:
        return 'General Fitness';
    }
  }
}

extension ExperienceLevelExtension on ExperienceLevel {
  String get displayName {
    switch (this) {
      case ExperienceLevel.beginner:
        return 'Beginner';
      case ExperienceLevel.intermediate:
        return 'Intermediate';
      case ExperienceLevel.advanced:
        return 'Advanced';
    }
  }
}

extension NutritionGoalExtension on NutritionGoal {
  String get displayName {
    switch (this) {
      case NutritionGoal.cut:
        return 'Cut (Lose Weight)';
      case NutritionGoal.bulk:
        return 'Bulk (Gain Weight)';
      case NutritionGoal.recomp:
        return 'Recomp (Body Recomposition)';
      case NutritionGoal.maintain:
        return 'Maintain Weight';
    }
  }
}

File: backend/app/models/user_profile_model.py
"""
User Profile Model
Stores extended user profile information separate from authentication
"""
from sqlalchemy import Column, Integer, String, Float, Enum as SQLEnum, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from ..core.database import Base
import enum
import uuid


class GenderEnum(str, enum.Enum):
    MALE = "male"
    FEMALE = "female"
    OTHER = "other"
    PREFER_NOT_TO_SAY = "prefer_not_to_say"


class FitnessGoalEnum(str, enum.Enum):
    STRENGTH = "strength"
    MUSCLE_GAIN = "muscle_gain"
    FAT_LOSS = "fat_loss"
    ENDURANCE = "endurance"
    GENERAL_FITNESS = "general_fitness"


class ExperienceLevelEnum(str, enum.Enum):
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


class NutritionGoalEnum(str, enum.Enum):
    CUT = "cut"
    BULK = "bulk"
    RECOMP = "recomp"
    MAINTAIN = "maintain"


class UserProfile(Base):
    __tablename__ = 'user_profiles'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id', ondelete='CASCADE'), unique=True, nullable=False)

    # Basic info
    full_name = Column(String, nullable=True)
    phone_number = Column(String, nullable=True)
    gender = Column(SQLEnum(GenderEnum, values_callable=lambda obj: [e.value for e in obj]), nullable=True)

    # Physical attributes
    weight_kg = Column(Float, nullable=True)
    height_cm = Column(Float, nullable=True)

    # Fitness preferences
    fitness_goal = Column(SQLEnum(FitnessGoalEnum, values_callable=lambda obj: [e.value for e in obj]), nullable=True)
    experience_level = Column(SQLEnum(ExperienceLevelEnum, values_callable=lambda obj: [e.value for e in obj]), nullable=True)
    training_frequency = Column(Integer, nullable=True)  # workouts per week
    nutrition_goal = Column(SQLEnum(NutritionGoalEnum, values_callable=lambda obj: [e.value for e in obj]), nullable=True)

    # Relationship
    user = relationship("User", back_populates="profile")
